### PR TITLE
feat(MinterGateway): require strict number of signatures for updateCollateral

### PR DESCRIPTION
**Pull request by PierrickGT**

### CODE REVIEW

File: src/MinterGateway.sol [hunk 1]
```diff
@@ -1051,20 +1051,21 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
+        for (uint256 index_; index_ < signatures_.length; ++index_) {
             unchecked {
                 // Check that validator address is unique and not accounted for
                 // NOTE: We revert here because this failure is entirely within the minter's control.
                 if (index_ > 0 && validators_[index_] <= validators_[index_ - 1]) revert InvalidSignatureOrder();
             }
 
             // Check that validator is approved by TTG.
-            if (!isValidatorApprovedByTTG(validators_[index_])) continue;
+            if (!isValidatorApprovedByTTG(validators_[index_])) revert NotApprovedValidator(validators_[index_]);
```
Comment (deluca-mike) [inline src/MinterGateway.sol:1068]:
```suggestion
            _revertIfNotApprovedValidator(validators_[index_]);
```

File: src/MinterGateway.sol [hunk 2]
```diff
@@ -1082,26 +1083,11 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
             );
 
             // Check that ECDSA or ERC1271 signatures for given digest are valid.
-            if (!SignatureChecker.isValidSignature(validators_[index_], digest_, signatures_[index_])) continue;
+            if (!SignatureChecker.isValidSignature(validators_[index_], digest_, signatures_[index_]))
+                revert InvalidSignature();
```
Comment (deluca-mike) [inline src/MinterGateway.sol:1087]:
```suggestion
            if (!SignatureChecker.isValidSignature(validators_[index_], digest_, signatures_[index_])) {
                revert InvalidSignature();
            }
```

**Review by deluca-mike — COMMENTED** [review]
While the code changes are good, it is unclear what this PR solves other than increasing the risk of a minter's collateral updates being invalidated while in flight due to TTG proposal executions. The code before this PR better meets the [Robustness Principle](https://en.wikipedia.org/wiki/Robustness_principle) since it is more liberal in what it accepts, yet equally conservative in what it does. Providing more than enough valid signatures from unique validators should never fail.

Now, if any of the following happens between the time of signature collection and transaction mining, the tx will fail:
- the threshold increases
- the threshold _decreases_
- alice (a used validator) is removed
- alice (a used validator) swaps key/address
- unknown state/logic in a validator ERC1271 contract changes

Prudent minters are likely to collect more valid signatures than needed for more transparency and more determination of the collateral success of the collateral updates. We should not discourage this, especially not if it also means being less robust.

File: src/MinterGateway.sol (hunk 1 above)
Comment (PierrickGT) [reply src/MinterGateway.sol]: Fixed in this commit: https://github.com/MZero-Labs/protocol/pull/165/commits/2671e1ed4b0491b1c3c102e73b86f4429b6b47f3

File: src/MinterGateway.sol [hunk 3]
```diff
@@ -1065,11 +1069,15 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
 
             // Check that validator is approved by TTG.
-            if (!isValidatorApprovedByTTG(validators_[index_])) revert NotApprovedValidator(validators_[index_]);
+            _revertIfNotApprovedValidator(validators_[index_]);
 
             // Check that the timestamp is not 0.
             if (timestamps_[index_] == 0) revert ZeroTimestamp();
 
+            // Verify that the validator's signature is newer than the previous one.
+            if (timestamps_[index_] < _lastValidatorUpdateTimestamp[minter_][validators_[index_]])
```
Comment (toninorair) [inline src/MinterGateway.sol:1078]: <=

**Comment by toninorair** [comment]
Originated this pr https://github.com/MZero-Labs/protocol/pull/166

**Pull request closed without merging**